vcd/resource_vcd_edgegateway.go

```diff
@@ -355,9 +504,309 @@ func resourceVcdEdgeGatewayDelete(d *schema.ResourceData, meta interface{}) erro
+		var subnetParticipation *types.SubnetParticipation
+		if defaultGatewayNetwork != "" && extNet.ExternalNetwork.Name == defaultGatewayNetwork {
+			for _, net := range extNet.ExternalNetwork.Configuration.IPScopes.IPScope {
+				if net.IsEnabled {
+					subnetParticipation = &types.SubnetParticipation{
+						Gateway: net.Gateway,
+						Netmask: net.Netmask,
+					}
+					break
+				}
+			}
+		}
+		networkConf := &types.GatewayInterface{
+			Name:          extNet.ExternalNetwork.Name,
+			DisplayName:   extNet.ExternalNetwork.Name,
+			InterfaceType: "uplink",
+			Network: &types.Reference{
+				HREF: extNet.ExternalNetwork.HREF,
+				ID:   extNet.ExternalNetwork.ID,
+				Type: "application/vnd.vmware.admin.network+xml",
+				Name: extNet.ExternalNetwork.Name,
+			},
+			UseForDefaultRoute:  defaultGatewayNetwork == extNet.ExternalNetwork.Name,
+			SubnetParticipation: []*types.SubnetParticipation{subnetParticipation},
+		}
+
+		gatewayInterfaces[extNetworkIndex] = networkConf
+	}
+	return gatewayInterfaces, nil
+}
+
+// getGatewayInterfacesType extracts `external_network` blocks with more advanced settings into
+// []*types.GatewayInterface
+// This is a pretty complicated function with 3 level nesting as it is implied by API structure.
+// This structure is documented below.
+//
+// <GatewayInterface>						<---- maps directly to `external_network` block
+// 	<Name>test_external_network</Name>
+// 	<DisplayName>test_external_network</DisplayName>
+// 	<Network href="...." id="urn:vcloud:network:144bafa4-7cbe-44de-a647-e6e045b7b8c5" type="application/vnd.vmware.admin.network+xml" name="test_external_network"></Network>
+// 	<InterfaceType>uplink</InterfaceType>
+// 	<SubnetParticipation>					<----- maps to nested `subnet` block(s) inside `external_network`
+// 		<Gateway>192.168.30.49</Gateway>
+// 		<Netmask>255.255.255.240</Netmask>
+// 		<IpAddress>192.168.30.51</IpAddress>
+// 		<IpRanges>							<----- maps to `suballocate_pool` block(s) inside `subnet`
+// 			<IpRange>
+// 				<StartAddress>192.168.30.53</StartAddress>
+// 				<EndAddress>192.168.30.55</EndAddress>
+// 			</IpRange>
+// 			<IpRange>
+// 				<StartAddress>192.168.30.58</StartAddress>
+// 				<EndAddress>192.168.30.60</EndAddress>
+// 			</IpRange>
+// 		</IpRanges>
+// 		<UseForDefaultRoute>true</UseForDefaultRoute>
+// 	</SubnetParticipation>
+// 	<SubnetParticipation>				    <---- simple `subnet` block without suballocated pools and automatic IP assignment
+// 		<Gateway>292.168.30.49</Gateway>
+// 		<Netmask>255.255.255.240</Netmask>
+// 		<UseForDefaultRoute>true</UseForDefaultRoute>
+// 	</SubnetParticipation>
+// 	<ApplyRateLimit>true</ApplyRateLimit>
+// 	<InRateLimit>100</InRateLimit>
+// 	<OutRateLimit>100</OutRateLimit>
+// 	<UseForDefaultRoute>true</UseForDefaultRoute>
+// </GatewayInterface>
+func getGatewayInterfacesType(vcdClient *VCDClient, externalInterfaceSet *schema.Set) ([]*types.GatewayInterface, error) {
```

Comment:
Can we divide this function with meaningful names to make self documented code?

Reply:
I split it, but IMHO the names are already meaningful as they can be with this API structure for interfaces. To me now it is even harder to follow it, but you can check if it is more readable for fresh eyes.